### ISSUE TITLE
Exclude our cache update javascript from popular Caching Plugins

### ISSUE THREAD

**Issue by Nicholas-Cardot** [warfare-plugins/social-warfare-pro#12]
Let's just try to start with W3 Total Cache and WP Super Cache

**Comment by nutsandbolts** [comment]
And WP Rocket

**Comment by Nicholas-Cardot** [comment]
Ideally, if we could simply exclude a small code snippet from being cached at all.

**Comment by Nicholas-Cardot** [comment]
Excess Ajax Calls Problems
-Social Warfare taxes server with excessive calls via ajax on every page load.
The ajax call is said to use a lot of cpu usage.
The ajax call only happens every page load if the Javascript for the call is cached
(solution: Excess Ajax Calls)

Excess Ajax Calls Solutions
-Solution could be to set a timestamp in the ajax call, and if beyond a few minutes, not make the update call.

**Comment by scott-wiser** [comment]
Here's my understanding right now (some items might not be needed, just noting them for good measure)


1) Have server drop an inline <script> tag that stores a global variable for server time when the page is prepared.

Either be sure to (1) have the inline script load when the problem script is loaded, or (2) have the inline script load at various point and contain a variable name specifying the point, hook, or time it loads

2) Have a JS function to compare browser time with server time (JS browser time might not always be reliable so consider better ways such as external sites or maybe even headers)

If the server date is too far back, stop the update, as it probably indicates a cached page was sent

3) Also consider a way to store last update time if it isn't already.  If it already is, then consider using that in the comparison somehow

4) Consider a variable to measure whether the PHP ob cacher was active

5) Consider a variable to measure whether a caching plugin is installed, activated, enabled, and operating during the time of the script

6) Have the variables in a global context, possible in a global array or object for organization, with the intent of using them elsewhere if possible and adding more variables.

So like a Social Warfare global array/object containing runtime relevant variables 

7) Consider a cookie or browser storage to measure for last update in the event of page reloads.  This would likely be more useful to prevent repeated updates on fast reloads though

**Comment by Nicholas-Cardot** [comment]
The simplest approach is often the best approach. In this instance, I want to remind you how I view this process. 

1. The PHP runs it's checks and decides that the cache is expired so it outputs some JS variable that essnetially ask the Javascript functions to go ahead and take care of business.
2. The javascript functions then fire off (including calls to admin-ajax.php).

So the simplest approach is to add a timestamp to the request the moment the request is being made. So the PHP will say, "Hey, Javascript, can you rebuild the cache? Oh, and by the way, I'm sending you this request at 7:03PM" The javascript sees the request, sees that it is currently 7:03PM and runs the function.

But the page gets cached so five minutes later when the page is loaded, Javascript gets a message from the server saying, "Hey, Javascript, can you rebuild the cache? Oh, and by the way, I'm sending you this request at 7:03PM" The javascript see the request, sees that it is currently 7:08PM, and says, "Nah. I already process that request 5 minutes ago.

So add a unix timestamp to the variables that get output with the cache rebuild request, once we are 60 seconds past that timestamp, start ignoring it forever. We won't run it again until the page has been rebuilt with a new request.  

While a few pages may receive multiple page loads during that first 60 seconds, that is still 1,000% fewer reduntant calls then just letting it keep running for ever. 

So, in conclusion:
1. Output a unix timestamp from PHP for javascript to see.
2. Make all js functions ignore the request if the timestamp is more than 60 seconds old.

**Comment by Nicholas-Cardot** [comment]
Right now, the JS puts out this:

```
    var swp_cache_data = {
	'action': 'swp_cache_trigger',
	'post_id': <?php echo $info['postID']; ?>
    };
```

Instead, it should be putting out this:
```

    var swp_cache_data = {
	'action': 'swp_cache_trigger',
	'post_id': <?php echo $info['postID']; ?>,
        'timestamp':<?php echo $unixTimestamp; ?>
    };
```

Then mod the js to ignore the request if that timestamp is more than 60 seconds old.

**Comment by scott-wiser** [comment]
I'm working on this now.

**ob_start system usage**
I noted the ob_start(); is used for the JS output, which is then later placed into a $info['footer_output'] variable with ob_get_clean();

This is one valid way to write and store data.  My only concern is if another system is using the ob_start/ob_get_clean functions some conflicts might occur.  I'm not sure if this has happened.  I'll keep an eye on this while testing just in case.

It is possible to read the JS into a variable directly line by line or in a big chunk.

**Comment by scott-wiser** [comment]
**Time difference problem**
In my first tests, I'm finding that there is a 300 second, or 5 minute difference, between my browser and my online test server.  So the test for 60 seconds fails everytime, simply because of a time difference on the server and on my computer.

I suggest we consider a longer window of time to reduce time differences between the server and browser.

I'll finish up and set it to 60 for the commit.  To change this later, we just need to change a number and any messages.

**Comment by scott-wiser** [comment]
Update (Also noted in commit):
-Added timestamp output from server
-Added JS browser timestamp comparison
-Made ajax call for swp_cache_trigger only work if within timelimt
-Tested numerous times both above and below the time limit.
-Made Facebook check (socialWarfarePlugin.fetchShares) only work if within timelimit
--Had to move fetchShares call to be within document.onready because it needs to know if the time is within limits

FACEBOOK FETCH TEST INADEQUACY
-The fetchShares function did not output a message indicating a Facebook response.  This could be for a number of reasons, namely the
test environment is not yet configured with a Facebook account, and the test site/urls have 0 shares.  I recommend testing this change
on a page that has Facebook shares.

TIME DIFFERENCE PROBLEM
See previous GitHub message:
https://github.com/warfare-plugins/social-warfare-pro/issues/12
